controllers/controllers_test.go

```diff
@@ -0,0 +1,918 @@
+				APIVersion: corev1.SchemeGroupVersion.Version,
+				Kind:       "Namespace",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "test-reconcile-initialize",
+			},
+		}
+
+		g.Expect(fakeClient.Create(ctx, ns)).To(Succeed())
+
+		return ns
+	}
+
+	teardown := func(t *testing.T, g *WithT, ns *corev1.Namespace) {
+		t.Helper()
+
+		t.Log("Deleting the namespace")
+		g.Expect(fakeClient.Delete(ctx, ns)).To(Succeed())
+	}
+
+	g := NewWithT(t)
+	namespace := setup(t, g)
+	defer teardown(t, g, namespace)
+
+	cluster := newCluster(&types.NamespacedName{Name: "foo", Namespace: namespace.Name})
+	cluster.Spec = clusterv1.ClusterSpec{
+		ControlPlaneEndpoint: clusterv1.APIEndpoint{
+			Host: "test.local",
+			Port: 9999,
+		},
+	}
+	g.Expect(fakeClient.Create(ctx, cluster)).To(Succeed())
+	patchHelper, err := patch.NewHelper(cluster, fakeClient)
+	g.Expect(err).To(BeNil())
+	cluster.Status = clusterv1.ClusterStatus{InfrastructureReady: true}
+	g.Expect(patchHelper.Patch(ctx, cluster)).To(Succeed())
+
+	genericInfrastructureMachineTemplate := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"kind":       "GenericInfrastructureMachineTemplate",
+			"apiVersion": "infrastructure.cluster.x-k8s.io/v1beta1",
+			"metadata": map[string]interface{}{
+				"name":      "infra-foo",
+				"namespace": cluster.Namespace,
+			},
+			"spec": map[string]interface{}{
+				"template": map[string]interface{}{
+					"spec": map[string]interface{}{
+						"hello": "world",
+					},
+				},
+			},
+		},
+	}
+	g.Expect(fakeClient.Create(ctx, genericInfrastructureMachineTemplate)).To(Succeed())
+
+	tcp := &controlplanev1.TalosControlPlane{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: cluster.Namespace,
+			Name:      "foo",
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					Kind:       "Cluster",
+					APIVersion: clusterv1.GroupVersion.String(),
+					Name:       cluster.Name,
+					UID:        cluster.UID,
+				},
+			},
+		},
+		Spec: controlplanev1.TalosControlPlaneSpec{
+			Replicas: nil,
+			Version:  "v1.16.6",
+			InfrastructureTemplate: corev1.ObjectReference{
+				Kind:       genericInfrastructureMachineTemplate.GetKind(),
+				APIVersion: genericInfrastructureMachineTemplate.GetAPIVersion(),
+				Name:       genericInfrastructureMachineTemplate.GetName(),
+				Namespace:  cluster.Namespace,
+			},
+			ControlPlaneConfig: controlplanev1.ControlPlaneConfig{},
+			RolloutStrategy: &controlplanev1.RolloutStrategy{
+				Type: "RollingUpdate",
+				RollingUpdate: &controlplanev1.RollingUpdate{
+					MaxSurge: &intstr.IntOrString{
+						IntVal: 1,
+					},
+				},
+			},
+		},
+	}
+	g.Expect(fakeClient.Create(ctx, tcp)).To(Succeed())
+
+	corednsCM := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "coredns",
+			Namespace: namespace.Name,
+		},
+		Data: map[string]string{
+			"Corefile": "original-core-file",
+		},
+	}
+	g.Expect(fakeClient.Create(ctx, corednsCM)).To(Succeed())
+
+	kubeadmCM := &corev1.ConfigMap{
```

Comment:
what is that???

Reply:
copy-paste issue ;) This PR is still WIP. Should have marked it as draft, debugging `unit-tests` step in CI